core/src/main/java/hudson/PluginManager.java

```diff
@@ -1367,7 +1367,7 @@ public HttpResponse doPluginsSearch(@QueryParameter String query, @QueryParamete
                             .anyMatch(category -> StringUtils.containsIgnoreCase(category, query)) ||
                         plugin.hasWarnings() && query.equalsIgnoreCase("warning:");
                 })
-                .limit(limit)
+                .limit(Math.max(limit - plugins.size(), 1))
                 .sorted((o1, o2) -> {
```

Comment:
Is it weird that we're limiting the number of hits first, and sort by popularity later? Not directly related to this change (I think), but still. For a given popular term, I'd like to get the most relevant results from among all hits, not the first 20 or so hits in the list, sorted by relevance.

@timja Thoughts?

Reply:
And, the final result should perhaps be sorted as well? Not only per site.

Reply:
Right, I would expect this to aggregate one list across all update sites, sort that, and show the top N hits. Obviously if different sites use different popularity scales, that may be slightly weird, but it's probably the best we can do without giving every update sites its own `limit` entries.

So in a limited way, it is related to the PR :-)

Reply:
I think it was a performance issue, I tried that initially and it was super slow. 

But you could do the sorting once on load of the update center data rather than processing it each time.

or other solutions...